src/ORM/DataList.php

```diff
@@ -1398,6 +1421,33 @@ private function addEagerLoadedDataToParent(
         }
     }
 
+    /**
+     * NOTE: Do not change `DataList` to `static` in this method signature.
+     * Subclasses of DataList must still accept DataList arguments and return DataList!
+     */
+    private function manipulateEagerLoadingQuery(
+        DataList $fetchList,
+        string $relationChain,
+        string $relationType
+    ): DataList {
+        $filterCallback = $this->eagerLoadAllRelations[$relationChain];
+        if ($filterCallback !== null) {
+            $fetchList = $filterCallback($fetchList);
+        }
+        if (!($fetchList instanceof DataList)) {
+            throw new LogicException(
+                "Eagerloading callback for $relationType relation $relationChain must return a DataList."
+            );
+        }
+        $limit = $fetchList->dataQuery->getLimit();
```

Comment:
```suggestion
        $limit = $fetchList->dataQuery->query()->getLimit();
```
That's used in first() and last()

It will trigger DataQuery::getFinalisedQuery(), though the performance loss from that will be minuscule

Reply:
I really don't like this (it feels untidy), but given it's already used in this class for similar reasons it's not worth fighting about.

Done